### DISCUSSION TITLE
android armeabi-v7a build error.

### DISCUSSION THREAD

**Discussion by jianglei12138** [animetosho/md5-optimisation#8] [Q&A]
![image](https://github.com/animetosho/md5-optimisation/assets/17866650/980a1aac-f96e-4a26-b051-942580422146)

Hello Authors,
I got "inline assembly requires more registers than available" error when build for android armeabi-v7a.
how to fix?
and i have to say,  that's fantastic project

**Comment by jianglei12138** [comment]
Oh, I have been test with ndk 26.1.10909125(clang) and ndk r15c(gcc 4.9.x)

**Comment by animetosho** [comment]
Thanks for the interest in the code here.

The ARM code was something I started, but didn't really go through with here. It doesn't implement the optimisation tricks listed on the front page.  
This repository was mostly to serve as info and some x86 benchmarks to demo them, as opposed to code that one would actually use.

I did write "proper" [ARM](https://github.com/animetosho/par2cmdline-turbo/blob/turbo/parpar/hasher/md5-arm-asm.h) and [ARM64](https://github.com/animetosho/par2cmdline-turbo/blob/turbo/parpar/hasher/md5-arm64-asm.h) MD5 implementations for my ParPar application, but they aren't designed as a library, so you'd have to figure out how to make it work if you want to use it.  
I've also found that inline assembly can be finicky with register allocation; the [inclusion](https://github.com/animetosho/par2cmdline-turbo/blob/c32ff7b10c2d160820d69f75fc0b38ff44c8e8d2/parpar/hasher/md5-scalar.h#L17-L26) does avoid the ASM version if compiler optimisations aren't enabled.

Maybe I can export the MD5 engine from ParPar into a library if there's a lot of interest in it.  
(though it might make more sense to include the optimisations in something like OpenSSL, since that would affect more people)

**Comment by jianglei12138** [comment]
Thanks, I'll try parpar
